Dev tools: Improve developer UX for workflow building

thomas- opened the pull request:
This utilises the recently added O.currentAuthenticatedUser and enhancements to Impersonation endpoints (eg: redirects, impersonations while impersonating) to add tooling around workflow development:

* One click-impersonate: Quickly impersonate the current actionable user of a workflow from the action panel and return to the current page. (Where the actionable is a group, we get the first member of the group)
* Entity impersonate: Quickly impersonate any resolvable entity defined as part of a workflow with a sidebar on the right.
* Previous state: Quickly toggle back to the previous state from the workflow action panel

The tools are disabled by default but can be toggled on/off by a super user for their account from the 'Workflow admin' section.

We utilise the O.currentAuthenticatedUser API to allow debug/admin actions to be undertaken while impersonating if this is a development system (with plugin debugging enabled) and the user has enabled the tooling.

bensummers commented on std_workflow_dev/js/std_workflow_debug.js:
```diff
@@ -0,0 +1,131 @@
+        // --------------------------------------------------------------------------
+
+        P.respond("GET,POST", "/do/workflow-dev/debug/debug-mode", [
+            {pathElement:0, as:"string"},
```
Isn't this redundant now?

thomas- replied:
Can be used as non-auth user is the only main difference currently I think, *but* I want to/am imagining that the two diverge, where debug has more power and abilities that wouldn't be suitable for administrators even in the production system.

bensummers commented on std_document_store_dev/requirements.schema:
```diff
@@ -0,0 +1,3 @@
+
+
+
```
empty file!

bensummers commented on std_workflow_dev/plugin.json:
```diff
@@ -0,0 +1,17 @@
+  "displayDescription": "Development tools for workflow plugin",
+  "apiVersion": 4,
+  "depend": ["std_workflow"],
+  "privilegesRequired": ["pDatabase"],
```
Don't think it needs pDatabase?

bensummers commented on std_document_store_dev/plugin.json:
```diff
@@ -0,0 +1,15 @@
+  "pluginVersion": 1,
+  "displayName": "Document Store Development",
+  "displayDescription": "Development tools for document store plugin",
+  "installSecret": "02b34388d0f255452e4dbfa4ddea3946bd0b3205",
```
Standard plugins shouldn't have install secrets.

bensummers commented on std_document_store_dev/js/docstore_workflow_debug.js:
```diff
@@ -0,0 +1,136 @@
+        return (O.PLUGIN_DEBUGGING_ENABLED &&
+            O.currentAuthenticatedUser &&
+            O.currentAuthenticatedUser.isSuperUser &&
+            O.currentAuthenticatedUser.id === 3 && // SUPPORT only for now
```
Perhaps we should just let anyone use this? Some developers aren't going to be logging in via a dev portal which gets them SUPPORT.

thomas- replied:
Agreed. My intent here was being overly cautious until it's been used a bit and then dropping such a restriction.

bensummers commented on std_document_store_dev/js/docstore_workflow_debug.js:
```diff
@@ -0,0 +1,136 @@
+/* Haplo Platform                                     http://haplo.org
+ * (c) Haplo Services Ltd 2006 - 2016    http://www.haplo-services.com
```
It's 2017 & https: now...

bensummers commented on std_workflow_dev/js/std_workflow_debug.js:
```diff
@@ -0,0 +1,131 @@
+        var showDebugTools = function() {
+            return (O.PLUGIN_DEBUGGING_ENABLED &&
+                O.currentAuthenticatedUser &&
+                O.currentAuthenticatedUser.isSuperUser &&
```
This stuff is actually probably useful for support. Maybe these things do belong in the standard plugin?

thomas- replied:
Do you mean showing of admin tools while impersonating? Main worry for me is confusing users but maybe one for further down the line?

thomas- commented:
fixed the obvious issues that needed tidying, and added a commit that modifies std_workflow slightly so that we get a more usable admin tools when debugging
